File: java-ladidol/src/com/feng/draft/垃/L70爬楼梯.java
package com.feng.draft.垃;

import java.util.HashMap;
import java.util.Map;

/**
 * @Description:
 * @Author Ladidol
 * @Date: 2022/3/25 15:00
 * @Version 1.0
 */
public class L70爬楼梯 {
    public static void main(String[] args) {
        System.out.println("爬十阶楼梯有 " + new 爬楼梯().climbStairs(10));
    }
}
class 爬楼梯 {
    public Map<Integer, Integer> note = new HashMap<> ();
    public int climbStairs(int n) {
        if( n <= 2){
            return n;
        }
        //按这种,代表每一次分叉都会开启一次递归,有重复的递归
//        return climbStairs( n - 1) + climbStairs( n - 2);

        //按这种,代表每一个台阶都只开启一次递归
        if (note.containsKey(n)){
            return note.get(n);
        }else {
            note.put(n,climbStairs(n-1)+climbStairs(n-2));
            return note.get(n);
        }
    }
}

File: golang-devildyw/simple/L70爬楼梯.go
package main

func main() {

}

func climbStairs(n int) int {
	if n <= 2 {
		return n
	}
	dp := make([]int, n+1)
	dp[1] = 1
	dp[2] = 2
	for i := 3; i <= n; i++ {
		// 跳跃到第i个位置 dp[i-2] 到这个位置只有1种方法 跳两步 因为跳一步再跳一步已经被 dp[i-1]包含了 i-2的位置跨一步步就到i-1了嘛 已经包含了
		//dp[i-1] 到这个位置有1种方法
		dp[i] = dp[i-1] + dp[i-2]
	}
	return dp[n]
}
